### PR TITLE
Load save split directories 102

### PR DESCRIPTION
**Pull request by oliver-szymanski**
This enhancement #102 allows you to:

- save files as a directory structure instead of a cmpgn (zipped) file
; therefore you use git or another source control system to manage them and you can open/check files with an editor (only experts should do the latter)
- it splits the big content.xml file in multiple files per campaign/zone/token/macros; make it easier when using a source control system to detect/manage changes
- it puts the actual macro code in a ".macro" plain text file, so you can edit the code easily with any editor if you like
- allows to copy whole zones or just tokens or macros between campaigns; e.g. have a base map with all your macros and use that from a linked shared folder. so you can have lots of campaign sharing code

---
This change is [Reviewable](https://reviewable.io/reviews/rptools/maptool/371)

### CODE REVIEW

**Comment by oliver-szymanski** [comment]
I know I could have rewrote the whole load/save mechanism, but I tried to have as less impact as possible to have no issues in export/save/save as/autosave/backup. So this solution is based on the existing save/load mechanisms

**Comment by JamzTheMan** [comment]
@cwisniew Since we have an extra week (next release end of April?) do we want to merge this now and start reviewing/testing? Pending code review of course...

**Comment by Azhrei** [comment]
Since Craig implied that revert wasn’t as clean as he thought it’d be, should we create a new branch off develop to test such a big change?  It would make it easier to remove later (just delete the branch).  We can still tag it so the CIs generate builds...

**Comment by JamzTheMan** [comment]
Yea, good idea. I'll create a feature branch so we can at least run from source and test but not block/disrupt other 1.5.2 dev work.

@rkathey or @phergus Let me know if you need a RC for this but I think you guys can checkout the branch and run ok.

**Comment by oliver-szymanski** [comment]
> Reviewed 14 of 14 files at r1.
> Reviewable status: all files reviewed, 3 unresolved discussions (waiting on @Azhrei, @cwisniew, and @oliver-szymanski)
> 
> 
> src/main/java/net/rptools/maptool/client/AppConstants.java, line 67 at r1 (raw file):
> 
>   public static final String CAMPAIGN_FILE_EXTENSION = ".cmpgn";
>   public static final String CAMPAIGN_DIRECTIORY_EXTENSION = "";
> 
> Do we want to use a extension of some sort? It may be helpful to denote everything in the campaign dir is for save only (but I assume you ignore other content added, like if we dumped a pdf or something in the directory?)
> Oh, ok, I see later you have the file filter as .cmpgndir so probably should set it here and use the constants in MapToolFrame below...
> 
> src/main/java/net/rptools/maptool/client/ui/MapToolFrame.java, line 236 at r1 (raw file):
> 
>       new MTFileFilter("cmpgn", I18N.getText("file.ext.cmpgn"));
>   private final FileFilter campaignDirectoryFilter =
>       new MTFileFilter("cmpgndir", I18N.getText("file.ext.cmpgnDirectory"), true);
> 
> We should use the defined constants here...
> 
> src/main/java/net/rptools/maptool/util/PersistenceUtil.java, line 557 at r1 (raw file):
> 
>     fileName = fileName.replace("'", "_");
>     fileName = fileName.replace("#", "_");
>     return fileName;
> 
> We don't have something like this already? Maybe in the lib? Could be a single regular expression.  There could also be one in Apache lib?

No idea. If you know I'm happy to change :)

**Comment by Phergus** [comment]
@JamzTheMan No need for RC for initial testing.  Checkout and build from shell is fine.

**Comment by JamzTheMan** [comment]
In MapToolFrame you have the filter set/hard coded to cmpgndir.  Although, is there a reason to just not use .cmpgn? It's the same extension BUT it is a dir vs file so it shouldn't confuse anything.

The downside of dir is we can never to file association with it (unless we through a file inside the dir that we associate with?)  Not a major concern, just popped in my head.

**Comment by oliver-szymanski** [comment]
> In MapToolFrame you have the filter set/hard coded to cmpgndir.  Although, is there a reason to just not use .cmpgn? It's the same extension BUT it is a dir vs file so it shouldn't confuse anything.
> The downside of dir is we can never to file association with it (unless we through a file inside the dir that we associate with?)  Not a major concern, just popped in my head.

See answer in my response above. So basically the one in MTFileFilter is not used at all. I'm not sure if a suffix for directories is good or not, still a bit undecided. I like it on Mac, but there its much better supported to look as a file

**Comment by Azhrei** [comment]
TL;DR on my phone, but an extension for top-level dir would be good on Mac since we can later set a file association for it (we don’t do that now, but I had plans for it).  However, not sure what happens if that extension is then used for a regular file!  I’ll test and reply here.

**Comment by JamzTheMan** [comment]
> TL;DR on my phone, but an extension for top-level dir would be good on Mac since we can later set a file association for it (we don’t do that now, but I had plans for it). However, not sure what happens if that extension is then used for a regular file! I’ll test and reply here.

Yea, wasn't sure if association would work with directories?
If it does work, then technically we would want both files and dirs associated to MapTool and load appropriately anyway.

Javapackager supposedly supported associations but I couldn't figure it out (documentation sucks) and it may only work if you used ant config.  I was going to wait and see if the new jpackager supports it.

**Comment by Azhrei** [comment]
On the Mac at least, there’s a script we can supply that the packager will use when building.  The script can do the work.  Worst case, a similar PowerShell script on Windows and Bash script on Linux...
